src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs

```diff
@@ -216,51 +204,11 @@ internal void EndWrite(IAsyncResult asyncResult)
             }
         }
 
-        private byte[] InternalBuffer
-        {
-            get
-            {
-                return _internalBuffer;
-            }
-        }
-
-        private void EnsureInternalBufferSize(int newSize)
+        private void EnsureInternalBufferSize()
         {
-            if (_internalBuffer == null || _internalBuffer.Length < newSize)
+            if (_internalBuffer == null)
             {
-                bool wasPinnable = _internalBufferFromPinnableCache;
-                byte[] saved = _internalBuffer;
-
-                if (newSize <= PinnableReadBufferSize)
-                {
-                    if (PinnableBufferCacheEventSource.Log.IsEnabled())
-                    {
-                        PinnableBufferCacheEventSource.Log.DebugMessage2("In System.Net._SslStream.EnsureInternalBufferSize IS pinnable", this.GetHashCode(), newSize);
-                    }
-
-                    _internalBufferFromPinnableCache = true;
-                    _internalBuffer = s_PinnableReadBufferCache.AllocateBuffer();
-                }
-                else
-                {
-                    if (PinnableBufferCacheEventSource.Log.IsEnabled())
-                    {
-                        PinnableBufferCacheEventSource.Log.DebugMessage2("In System.Net._SslStream.EnsureInternalBufferSize NOT pinnable", this.GetHashCode(), newSize);
-                    }
-
-                    _internalBufferFromPinnableCache = false;
-                    _internalBuffer = new byte[newSize];
-                }
-
-                if (saved != null && _internalBufferCount != 0)
-                {
-                    Buffer.BlockCopy(saved, _internalOffset, _internalBuffer, 0, _internalBufferCount);
-                }
-
-                if (wasPinnable)
-                {
-                    s_PinnableReadBufferCache.FreeBuffer(saved);
-                }
+                _internalBuffer = ArrayPool<byte>.Shared.Rent(ReadBufferSize);
             }
             else if (_internalOffset > 0 && _internalBufferCount > 0)
```

Comment:
Can we ever get into the state where _internalBuffer != null but one or both of these is 0?  What happens in that case?  At the very least, we should assert here.

Reply:
If internal buffer != null, then we have a rented buffer

If the offset is > 0 and the count > 0 then we need to shift the data to the front, 
if the count is zero and the offset is > 0 then we just reset the offset to zero
if the offset is zero and the count is > 0 then we have data but it's at the front anyway

So all of the above cases could/will happen and they are all dealt with, or am I missing something?

Reply:
(1) When would we ever get into the case where count == 0 and internalBuffer != null?  Is that possible?  I don't think it is.  If not, we should assert it.  
(2) We don't seem to reset the _internalCount to 0 when we allocate a new buffer.  Perhaps this is guaranteed somewhere else?
(3) We should fix the name of the method, since it no longer has anything to do with size.

More than anything, I'm trying to understand the possible states we can get into here.  It wasn't obvious to me from the code.

Reply:
1. Yes an assert would be good for that, as it shouldn't happen anymore, it's due to me modifying this before the release method was baked in, it should never happen now.
2. Do you mean when we allocate a brand new buffer (eg buffer == null?) or when we allocate a new one and copy. In the first instance when we free it we set everything to zero, in the second case the copy should keep the same length.
3. Agreed

I will generally clean it up.

Reply:
Re 2 -- I meant buffer == null.  If we're assuming it's always 0 when we reach here, we should assert that.  But personally I'd just set it to 0 here.

Reply:
I think the assert is needed because it should never happen due to
Either the buffer is null because we have never allocated... or it is null because of

```
private void ReturnReadBufferIfEmpty()
{
    if (_internalBuffer != null && _decryptedBytesCount == 0 && _internalBufferCount == 0)
    {
        ArrayPool<byte>.Shared.Return(_internalBuffer);
       _internalBuffer = null;
       _internalBufferCount = 0;
       _internalOffset = 0;
       _decryptedBytesCount = 0;
       _decryptedBytesOffset = 0;
    }
}
```